### PR TITLE
Add `AzureBlobStorageConnector`

### LINKED ISSUE #257: Create a data conector for Azure Blob Storage
Create a data connector that pulls data from [Azure Blob Storage ](https://learn.microsoft.com/en-us/azure/storage/blobs/)

See [Adding Data Connectors](https://github.com/Unstructured-IO/unstructured/blob/main/Ingest.md#adding-data-connectors) for details on how to get started. Make sure to include a link to this issue when submitting a PR.

**Definition of Done**

- The [checklist](https://github.com/Unstructured-IO/unstructured/blob/main/Ingest.md#the-checklist) has been completed.
- The connector is able to process a single object or a container.
- The connector is able to process the entire container.
- For private files, it requires the credentials of access.
- For now, it is OK to process only doc types that [partition.auto()](https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/partition/auto.py) is capable of processing. 
    
Bonus points: the ability to filter by document type.

### PR DESCRIPTION
**Pull request by alvarobartt**
## What's in this PR?

Upon #318 merge into `main`, this PR introduces the `AzureBlobStorageConnector` explicit `fsspec` connector so as to close #257. This PR also introduces a feature in the `ingest-cli` so as to use the generic `FsspecConnector` in case the protocol is not among the specifically implemented ones. Finally, a deprecation lifecycle starts for the `--s3-url` option in favor of `--remote-url`.

## What's missing in this PR?

Adding the `test_unstructured_ingest/test-ingest-azure.sh` script for testing, but it's still pending since the Azure Blob Storage container provided by @benjats07 in #257 contains some files that `unstructured` cannot parse/partition, in this case both XLSX and XML files.

### CODE REVIEW

**Comment by alvarobartt** [comment]
Hi @cragwolfe, this should also be done already 👍🏻 Feel free to install it as `pip install -e ".[azure,local-inference]" and run it as `./examples/azure/ingest.sh` or using the `unstructured-ingest` CLI as `unstructured-ingest --remote-url abfs://demo-files/ --azure-account-name datalakebenja22 --structured-output-dir azure-demo-files-output --num-processes 2`

File: unstructured/ingest/main.py
Comment (cragwolfe) [inline unstructured/ingest/main.py:109]: Might as well explicitly mention the supported protocols here, to save the user any confusion.

File: unstructured/ingest/main.py
Comment (cragwolfe) [inline unstructured/ingest/main.py:370]: readability nit: top 8 lines are a bit awkward. break into an old fashioned if/else on line 360?

```
    if azure_connection_string:
       access_kwargs = {"connection_string": azure_connection_string}
    elif azure_account_name:
       ...
```

**Comment by cragwolfe** [comment]
> Adding the test_unstructured_ingest/test-ingest-azure.sh script for testing, but it's still pending since the Azure Blob Storage container provided by @benjats07 in https://github.com/Unstructured-IO/unstructured/issues/257 contains some files that unstructured cannot parse/partition, in this case both XLSX and XML files.

You can now use:
```
        --remote-url abfs://container1/ \
        --azure-account-name azureunstructured1 \
```

**Comment by cragwolfe** [comment]
Besides the minor comments and once the tests are in CI, should be good to go!

**Comment by alvarobartt** [comment]
> > Adding the test_unstructured_ingest/test-ingest-azure.sh script for testing, but it's still pending since the Azure Blob Storage container provided by @benjats07 in #257 contains some files that unstructured cannot parse/partition, in this case both XLSX and XML files.
> 
> You can now use:
> 
> ```
>         --remote-url abfs://container1/ \
>         --azure-account-name azureunstructured1 \
> ```

Sure I'll update it now and retry to pass the CI :+1:

File: unstructured/ingest/main.py
Comment (tomaarsen) [inline unstructured/ingest/main.py:370]: This `AzureBlobStorageConnector` will, as it is simply an empty subclass of `FsspecConnector`, make instances of `FsspecIngestDoc`, rather than of `AzureBlobStorageIngestDoc`. As a result, the latter class is never used, and its `@requires_dependencies(["adlfs", "fsspec"], extras="azure")` is never called.
This last bit is not really a problem, as there is also a `requires_dependencies` decorator around `AzureBlobStorageConnector`.

Comment (alvarobartt) [reply unstructured/ingest/main.py:370]: Hmm true, maybe we can parametrize that to provide the `...IngestDoc` class to the `FsspecConnector` so that it's used instead of the default `FsspecIngestDoc`, or maybe we can just skip it and remove both `S3IngestDoc` and `AzureBlobStorageIngestDoc`?

File: unstructured/ingest/main.py
Comment (alvarobartt) [reply unstructured/ingest/main.py:369]: You're right! Ok I'll commit this manually and add you as co-author 👍🏻

File: unstructured/ingest/main.py
Comment (tomaarsen) [reply unstructured/ingest/main.py:370]: I think both solutions are reasonable. The first one would be as simple as:
```python
# azure.py
@requires_dependencies(["adlfs", "fsspec"], extras="azure")
class AzureBlobStorageConnector(FsspecConnector):
    ingest_doc_class = AzureBlobStorageIngestDoc
```

```python
# ffspec.py
class FsspecConnector(BaseConnector):
    """Objects of this class support fetching document(s) from"""

    ingest_doc_class = FsspecIngestDoc

...

    def get_ingest_docs(self):
        return [
            self.ingest_doc_class(
                self.config,
                file,
            )
            for file in self._list_files()
        ]
```

```python
# s3.py
@requires_dependencies(["s3fs", "fsspec"], extras="s3")
class S3Connector(FsspecConnector):
    ingest_doc_class = S3IngestDoc
```

And the second one is as simple as just removing the ingest classes from existence. Perhaps the first approach allows for more flexibility in case there ever comes an `ffspec`-supported ingest connector that wants to do some additional processing? If that never happens though, then the second approach is simpler. 
I think it's best to defer to the maintainers here.

**Comment by tomaarsen** [comment]
The logs are showing:
```
./test_unstructured_ingest/test-ingest.sh: line 9: ./test_unstructured_ingest/test-ingest-azure.sh: Permission denied
```

You likely have to set the permissions of this file such that they can be executed:
```
git update-index --chmod=+x ./test_unstructured_ingest/test-ingest-azure.sh
```
I ran into this issue myself with several of my ingest data connectors.

File: unstructured/ingest/main.py
Comment (alvarobartt) [reply unstructured/ingest/main.py:370]: So what I did is basically the same as you mentioned above but using `ingest_doc_cls` as an attribute 😆 But maybe makes more sense to just remove it and rely on `FsspecIngestDoc`?

**Comment by alvarobartt** [comment]
> The logs are showing:
> 
> ```
> ./test_unstructured_ingest/test-ingest.sh: line 9: ./test_unstructured_ingest/test-ingest-azure.sh: Permission denied
> ```
> 
> You likely have to set the permissions of this file such that they can be executed:
> 
> ```
> git update-index --chmod=+x ./test_unstructured_ingest/test-ingest-azure.sh
> ```
> 
> I ran into this issue myself with several of my ingest data connectors.

Hmm looks like this didn't work @tomaarsen I'll look for a solution! 👍🏻

**Comment by alvarobartt** [comment]
So it worked already... but it seems that we're getting limitations from GitHub, which is not related to this PR 😖 Can we just re-run those failing tests? @cragwolfe

**Comment by tomaarsen** [comment]
I suspect that all tests would pass if it wasn't for the rate limit on the GitHub tests, considering that the tests pass for Python 3.10 and the recent changes since the last passing tests should not have introduced any version-specific code. However, safest is just to wait a bit and rerun the tests.

**Review by cragwolfe — APPROVED** [review]
Nice! 🚀

**Pull request merged**
